lib/recommender_system/irt_recommender.rb

```diff
@@ -0,0 +1,193 @@
+module RecommenderSystem
+  class IrtRecommender < RecommenderSystem::Recommender
+    # Metoda vytvori list learning objektov s priradenymi hodnotami na zaklade pravdepodobnosti spravnej odpovede urcenej cez Item Response Theory
+    def get_list
+      los = self.learning_objects
+      list = Hash.new
+      user = User.find(self.user_id)
+      proficiency = user.proficiency
+      los.each do |lo|
+        difficulty = lo.irt_difficulty || 0.5
+        discrimination = lo.irt_discrimination || 0.5
+        # Vypocet pravdepodobnosti na zaklade 2P IRT
+        probability = (Math::E**(discrimination*(proficiency-difficulty))) / (1 + Math::E**(discrimination*(proficiency-difficulty)))
+        list[lo.id] = probability
+      end
+      list
+    end
+
+    # Opakuje sa o stvrtej rano, vid. config/schedule.rb
+    def self.update_table
+      require "rinruby"
+      # Predspracovanie dat
+      result_relations = Levels::Preproces.preproces_data
+      number_of_users = User.all.order("id DESC").first.id
+      number_of_los = LearningObject.all.order("id DESC").first.id
```

Comment:
Pocet objektov nie je najvyssie ID! ID v databaze ani nahodou nemusia ist za sebou po jednom. Okrem trochu exotickejsich pripadov:
- niekto sa pozrie na hodnotu DB sekvencie a zavola `nextval()`, dalsi vlozeny zaznam bude mat ID+2 miesto +1
- napr. Rails metoda `find_or_create` nie je atomicka a ak sa zavola z viacerych procesov sucasne, ID sa pokusia v prikaze `INSERT` ziskat oba, ale prebehne nakoniec len jeden, takze ID mozu byt povynechavane

Tak existuje predsa uplne zakladny pripad, ze niekto jednoducho pouzije DELETE a hociktory existujuci zaznam vymaze. ID nie je poradove cislo riadku, ze by sa posunuli riadky, to je kluc, ktory ked uz raz ma hodnotu, stale identifikuje rovnaky zaznam (predmet Databazove systemy).

A napriklad teraz na produkcii:
`SELECT MAX(id) FROM learning_objects;` => 677 (to iste trochu inak zisti ta kombinacia s all,order,first,id, stacilo by aj `LearningObject.maximum(:id)`)
ale:
`SELECT COUNT(id) FROM learning_objects;` => 676

Najvyssie ID je 677, ale zaznamov je len 676. Napriklad aktualne neexistuje zaznam s ID 157.

Ak to maju byt len maximalne rozmery matice nizsie, tak ok, ostane par stlpcov nevyuzitych (len potom by sa to nemalo volat "pocet lo", ale "najvacsie id lo"), lenze este nizsie sa to dava naspat do databazy cez cyklus s `i` od 1 po N a caka sa, ze LO bude mat ID o hodnote toho `i`. To je velmi zle. Asi to v pripade toho chybajuceho ID nateraz zachranuje to, ze sa to v tom cykle este prefiltruje cez `items_deleted` (co asi pozera, ci k LO existuje aspon 10 odpovedi, ziskanych z matice a cez metodu `preprocess_data`, ktora ale pozera user_to_lo_relations). Tu je ale dalsi problem, tak ako moze byt LO vymazane naozaj, moze byt skryte v DB cez scope (`act_as_paranoid` z gemu `paranoia`) - ak ho chceme "deaktivovat" pred zobrazovanim studentom, ale nechceme ho natvrdo vylucit navzdy z DB a prist tym o vsetky logy aktivity, spatnej vazby, atd. s nim suvisiace - a potom `LearningObject.find(i)` ho po spravnosti nenajde (lebo scope), ale user_to_lo_relations k nemu existuju, takze sa tu ani nevyfiltruje.
Toto by nebol problem, keby sa s ID pracuje tak, ako sa ma - proste je to mnozina/array IDciek (takze sa spytam, ake LO existuju a potom ich mozem iterovat), urcite neiterovat cisla od 1 do N a cakat, ze to budu vsetko IDcka.

Ak sa este hocikde inde v kode pracuje s tym, ze ID = 1..max(ID), treba to fixnut, lebo to moze viest k vela neocakavanym situaciam.